File: js/interviewbit/strings/AddBinaryStrings.js
module.exports = {
    //param A : string
    //param B : string
    //return a strings
    addBinary : function(A, B){
        var carry = 0;
        var res = "";
        var indexA = A.length - 1;
        var indexB = B.length - 1;
        while (indexA >= 0 || indexB >= 0 || carry != 0) {
            var current = carry;
            if (indexA >= 0) current += A[indexA--] - '0';
            if (indexB >= 0) current += B[indexB--] - '0';
            carry = current >> 1;
            current = current & 1;
            res = (current == 1 ? '1' : '0') + res;
        }
        return res;
    }
};

File: c++/Interviewbit/Strings/AddBinaryStrings.cpp
//
//  AddBinaryStrings.cpp
//  Interviewbit
//
//  Created by Shevchyk Dmytro on 04.08.2018.
//  Copyright © 2018 Shevchyk Dmytro. All rights reserved.
//

#include "AddBinaryStrings.hpp"

using std::to_string;

string Solution::addBinary(string A, string B) {
    int carry = 0;
    string res;
    auto currentA = A.rbegin();
    auto currentB = B.rbegin();
    while (currentA != A.rend() || currentB != B.rend() || carry != 0) {
        int current = carry;
        if (currentA != A.rend()) current += *(currentA++) - '0';
        if (currentB != B.rend()) current += *(currentB++) - '0';
        res = to_string(current % 2) + res;
        carry = current / 2;
    }
    return res;
}
